File: src/token.ts
export const enum TokenType {
  OP_ADD,
  OP_SUB,
  OP_MUL,
  OP_DIV,
  OP_NEG,
  OP_POS,
  OP_INC,
  OP_DEC,
  OP_EQ,
  COMP_EQ,
  COMP_NE,
  COMP_GT,
  COMP_GE,
  COMP_LT,
  COMP_LE,
  COMP_AND,
  COMP_OR,
  LOG_NOT,
  LOG_AND,
  LOG_OR,
  LOG_XOR,
  LITERAL,
  LPAREN,
  RPAREN,
  LBRACE,
  RBRACE,
  COLON,
  VAR_DECL,
  VAR,
  IN,
  OUT,
  IF,
  ELSE,
  WHILE,
  LABEL,
  GOTO,
  EXIT,
  FUNC_DECL,
  CALL,
  RETURN,
  IDENT,
  EVAL,
  EOF,
}

export type Literal = string | number | boolean | void;

export type Token = {
  type: TokenType;
  name?: string;
  value?: Literal;
};

File: src/token.rs
use ordered_float::OrderedFloat;

use crate::ast::Expr;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum TokenType {
    ADD,
    SUB,
    MUL,
    DIV,
    NEG,
    EQ,
    ADDEQ,
    SUBEQ,
    MULEQ,
    DIVEQ,
    COMPEQ,
    COMPNE,
    COMPGT,
    COMPGE,
    COMPLT,
    COMPLE,
    COMPAND,
    COMPOR,
    LOGNOT,
    LOGAND,
    LOGOR,
    LOGXOR,
    LITERAL(VarType),
    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,
    LBRACKET,
    RBRACKET,
    COLON,
    VARDECL,
    VAR,
    OUT,
    IF,
    ELSE,
    WHILE,
    FOR,
    IN,
    LABEL,
    GOTO,
    FUNCDECL,
    CALL,
    RETURN,
    IDENT,
    EXTERN,
    PUB,
    Type(VarType),
    SIZEOF,
    RANGE,
    COMMA,
    EOF,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Literal {
    Int(i64),
    Float(OrderedFloat<f64>),
    Bool(bool),
    Str(String),
    Array(usize, Vec<Expr>),
    Void,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum VarType {
    Int,
    Float,
    Bool,
    Str,
    Array(Option<usize>),
    Void,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token: TokenType,
    pub value: Option<Literal>,
    pub row: usize,
    pub col: usize,
}
